controllers/topology/internal/extensions/patches/inline/json_patch_generator.go

```diff
@@ -200,45 +200,110 @@ func calculateValue(patch clusterv1.JSONPatch, variables map[string]apiextension
+	// If the variable is a nested variable we inspect the variable object.
 
-	// Get the builtin variable object.
-	builtinsValue, ok := variables[patchvariables.BuiltinsName]
+	// Get the variable object.
+	value, ok := variables[strings.Split(variableName, ".")[0]]
 	if !ok {
-		return nil, errors.Errorf("variable %q does not exist", patchvariables.BuiltinsName)
+		return nil, errors.Errorf("variable %q does not exist", variableName)
 	}
 
-	// Parse the builtin variable object.
-	builtins, err := fastjson.ParseBytes(builtinsValue.Raw)
+	// Parse the variable object.
+	variable, err := fastjson.ParseBytes(value.Raw)
 	if err != nil {
-		return nil, errors.Errorf("cannot parse variable %q", patchvariables.BuiltinsName)
+		return nil, errors.Errorf("cannot parse variable %q", variableName)
 	}
 
-	// Split the variable name and exclude the first part ("builtins.")
+	// Split the variable name and exclude the first part ("<variableName>.")
 	relativePath := strings.Split(variableName, ".")[1:]
 
-	// Return if the builtin variable does not exist.
-	if !builtins.Exists(relativePath...) {
-		return nil, errors.Errorf("variable %q does not exist", variableName)
-	}
+	// Split the path and retrieve the variable value incrementally.
+	for _, p := range relativePath {
+		// Parse the path segment.
+		pathSegment, err := parsePathSegment(p)
+		if err != nil {
+			return nil, errors.Wrapf(err, "variable name %q has invalid syntax", variableName)
+		}
+
+		// Return if the variable does not exist.
+		if !variable.Exists(pathSegment.path) {
+			return nil, errors.Errorf("variable %q does not exist", variableName)
+		}
+
+		// Just get the variable from the variable object,
+		// if the path doesn't contain an index.
+		if pathSegment.index == -1 {
+			// arrayIndex not set.
+			variable = variable.Get(pathSegment.path)
+			continue
+		}
+
+		// Get the variable from the variable object
+		// by retrieving the correct array element.
+		arr, err := variable.Get(pathSegment.path).Array()
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get array %q", variableName)
+		}
 
-	// Get the builtin variable from the builtin variable object.
-	variableValue := builtins.Get(relativePath...)
+		if len(arr) < pathSegment.index+1 {
+			return nil, errors.Errorf("array does not have index %d", pathSegment.index)
+		}
+
+		variable = arr[pathSegment.index]
+	}
 
 	// Return the marshalled value of the variable.
 	return &apiextensionsv1.JSON{
-		Raw: variableValue.MarshalTo([]byte{}),
+		Raw: variable.MarshalTo([]byte{}),
+	}, nil
+}
+
+type pathSegment struct {
+	path  string
+	index int
+}
+
+func parsePathSegment(segment string) (*pathSegment, error) {
+	if (strings.Contains(segment, leftArrayDelim) && !strings.Contains(segment, rightArrayDelim)) ||
+		(!strings.Contains(segment, leftArrayDelim) && strings.Contains(segment, rightArrayDelim)) {
+		return nil, errors.Errorf("failed to parse path segment")
```

Comment:
Let's return the offending segment in case of errors
```suggestion
		return nil, errors.Errorf("failed to parse path segment %s", segement)
```
Same below